cdc/generation.cc

```diff
@@ -173,118 +167,34 @@ topology_description generate_topology_description(
-                --last_set_stream_idx;
-            }
-
-            if (last_set_stream_idx == -1) {
+        static thread_local std::mt19937_64 rand_gen(std::random_device().operator()());
+        static thread_local std::uniform_int_distribution<int64_t> rand_dist(std::numeric_limits<int64_t>::min());
+        dht::sharder sharder(entry.streams.size(), entry.sharding_ignore_msb);
+        int64_t next_token_for_unallocated_shard = dht::token::to_int64(prev_token) + 1;
+        int64_t upper_bound = dht::token::to_int64(entry.token_range_end);
+        for (size_t shard_idx = 0; shard_idx < entry.streams.size(); ++shard_idx) {
+            int64_t token;
+            auto t = sharder.token_for_next_shard(prev_token, shard_idx);
+            if (t > entry.token_range_end) {
                 cdc_log.warn(
-                        "CDC wasn't able to generate any stream for vnode ({}, {}]. We'll use another vnode's streams"
-                        " instead. This might lead to inconsistencies.",
-                        tokens[(idx + off + entries.size() - 1) % entries.size()], tokens[(idx + off) % entries.size()]);
-
-                ss[0] = some_stream;
-                last_set_stream_idx = 0;
-            }
-
-            some_stream = ss[last_set_stream_idx];
-
-            // Replace 'unset' stream ids with indexes below last_set_stream_idx
-            for (int s_idx = last_set_stream_idx - 1; s_idx > -1; --s_idx) {
-                if (ss[s_idx].is_set()) {
-                    some_stream = ss[s_idx];
-                } else {
-                    ss[s_idx] = some_stream;
-                }
-            }
-            // Replace 'unset' stream ids with indexes above last_set_stream_idx
-            for (int s_idx = ss.size() - 1; s_idx > last_set_stream_idx; --s_idx) {
-                if (ss[s_idx].is_set()) {
-                    some_stream = ss[s_idx];
+                        "CDC wasn't able to generate any stream on shard {} on vnode ({}, {}]. "
+                        "This shard will have its stream located on another shard in this vnode instead. "
+                        "This might lead to worse performance.",
+                        shard_idx, prev_token, entry.token_range_end);
+                token = next_token_for_unallocated_shard;
+                if (next_token_for_unallocated_shard == upper_bound) {
+                    next_token_for_unallocated_shard = dht::token::to_int64(prev_token) + 1;
+                } else if (next_token_for_unallocated_shard == std::numeric_limits<int64_t>::max()){
+                    next_token_for_unallocated_shard = std::numeric_limits<int64_t>::min();
                 } else {
-                    ss[s_idx] = some_stream;
+                    ++next_token_for_unallocated_shard;
                 }
+            } else {
+                token = dht::token::to_int64(t);
             }
+            entry.streams[shard_idx] = {token, rand_dist(rand_gen)};
```

Comment:
Continuing Avi's question above: Why do we need this random number here? Why can't we just take "token"?
This is also related to my question on the previous patch - why do we need this random second half of the stream id if anyway the token is just the first part? Can't we just lose the second part? Or if we can't, what does its randomness achieve?

Reply:
There are multiple generations of CDC streams. Every time a topology change appears the new generation is created. There can be two streams, one is old generation and another in new generation, that share the same token but we want their stream_ids to be different because those are different streams. They just happen to be represented by the same token. Does that make sense @nyh?

Reply:
I don't think I understand. If the topology change did not involve a specific vnode (changed neither of its ends), then we might very well continue the same stream, and don't need a new one. If it *did* involve a specific vnode, then if it changed the beginning of the range it will change the begin token itself - and if it changed the end of the range, isn't it fine that one stream will remain with the same name, and the second half of the vnode will get a new stream?
Even if this second part is necessary, if the second half is about a "generation" number, why not use it directly as the second half - not a random number (which, thanks to the birthday paradox, can get a collision more frequently than you might think).

Reply:
When the topology changes, we have to close all the streams and open a set of new streams that reflect new topology. This is because the client has to read all the records that appeared before the topology change prior to reading records that appear after the change. It has to be a clear cut if we want to maintain the order of the messages.

Do you really think that the chance for the collision is so big?

Reply:
If we create 1 billion CDC generations such that there will be some token that ends a vnode in all of those generations, the probability of collision for this token will be 0.03.